flasgger/utils.py

```diff
@@ -337,22 +343,90 @@ def validate(
 
-    for key, value in definitions.items():
-        if 'id' in value:
-            del value['id']
+    strict_validation = True
+    if request.method == 'GET':
+        strict_validation = False
+    validate_data(data, main_def, strict_validation=strict_validation)
+
+    # do not validate headers if data is not None
+    if not should_validate_headers:
+        return
+
+    # Headers validation
+
+    schema_id = schema_id_for_source('headers', params)
+
+    if not schema_id:
+        return
+
+    for defi in raw_definitions:
+        if defi['id'].lower() == schema_id.lower():
+            main_def = defi.copy()
+
+    data = {k.lower(): v for k, v in dict(request.headers).items()}
+    validate_data(data, main_def, strict_validation=False)
 
+
+def schema_id_for_source(source, params):
+    schema_id = None
+    for param in params:
+        if param.get('in') == source:
+            schema_id = param.get('schema', {}).get('$ref')
+            if schema_id:
+                schema_id = schema_id.split('/')[-1]
+                break  # consider only the first
+    return schema_id
+
+
+def validate_data(data, definition, validation_function=None, validation_error_handler=None,
+                  strict_validation=True):
     if validation_function is None:
-        validation_function = jsonschema.validate
+        if not strict_validation:
+            validation_function = liberal_validator().validate
+        else:
+            validation_function = jsonschema.validate
 
     try:
-        validation_function(data, main_def)
+        validation_function(data, definition)
     except Exception as err:
         if validation_error_handler is not None:
-            validation_error_handler(err, data, main_def)
+            validation_error_handler(err, data, definition)
         else:
             abort(Response(str(err), status=400))
 
 
+def liberal_validator():
+    """
+    Validator that returns true for integers or numbers inside string.
+    Useful for Headers and GET call validation.
+    :return: custom validator
+    """
+    def type_or_string_type(type_name):
+        type_cast = int
+        if type_name == 'number':
+            type_cast = float
+
+        def type_or_string_type(checker, instance):
+            if Draft4Validator.TYPE_CHECKER.is_type(instance, type_name):
+                return True
+
+            if checker.is_type(instance, "string"):
+                try:
+                    type_cast(instance)
+                    return True
+                except ValueError:
+                    pass
+            return False
+
+        return type_or_string_type
+
+    type_checker = Draft4Validator.TYPE_CHECKER.redefine('integer', type_or_string_type('integer'))
+    CustomValidator = extend(Draft4Validator, type_checker=type_checker)
+    type_checker = CustomValidator.TYPE_CHECKER.redefine('number', type_or_string_type('number'))
+    CustomValidator = extend(CustomValidator, type_checker=type_checker)
```

Comment:
Instead of doing this twice, theres a method which lets you add a mapping of redefined types